terrestris/shogun
# Replace deprecated keycloak-starter with oauth2-resource-server

## Description



This suggests to replace the keycloak-starter (that will [reach the end](https://www.keycloak.org/2022/02/adapter-deprecation) of support by the end of 2022) with oauth2-resource-server for authentication against the Keycloak instance.

Well, at first glance the change looked quite manageable, but during the update it turns out, that we could improve the overall integration of Keycloak a lot - especially in terms of an easier communication between all components - by following the OpenID approach (see sketch below) and making use a bearer token based authentication using [JWT tokens](https://jwt.io/). Thus a stateless session management is proposed which makes redis no longer necessary.

![auth-workflow](https://user-images.githubusercontent.com/1137620/175568561-0321d79f-4fc9-411a-811c-61e971f93c79.png)

HttpBasic authentication and a form login aren't supported anymore.

In addition this removes the static shogun-baseclient resources from the build artefact as the client (or better the [shogun-demo-client](https://github.com/terrestris/shogun-demo-client)) should be of preferred use and fixes the development setup for the interceptor module.

TODOs:

- The graphiql package/version we're currently using has no support for sending an auth token. This should be fixed in a follow up PR. 

## Migration notes:

- Applications updating to latest version of SHOGun might update the following entries in the respective `application.yml` (see [PR 43](https://github.com/terrestris/shogun-docker/pull/43) as reference):
  - Disable redis:
  ```yaml
  spring:
    session:
      store-type: none
  ```
  - Remove the `keycloakauth` key
  - Add the spring-resource config:
  ```yaml
  spring:
    security:
      oauth2:
        resourceserver:
          jwt:
            issuer-uri: https://localhost/auth/realms/SHOGun
            jwk-set-uri: https://localhost/auth/realms/SHOGun/protocol/openid-connect/certs
  ```
  - Update the `keycloak` key:
  ```yaml
  keycloak:
    enabled: true
    server-url: https://localhost/auth
    username: admin
    password: shogun
    master-realm: master
    admin-client-id: admin-cli
    realm: SHOGun
    client-id: shogun-boot
    principal-attribute: preferred_username
    disableHostnameVerification: true
  ```
  - Update the keycloak configuration as seen in [PR 42](https://github.com/terrestris/shogun-docker/pull/42).
  - Update the keycloak configuration for the GeoServer as seen in [PR 41](https://github.com/terrestris/shogun-docker/pull/41)
  - Update the client configuration to make use of the [keycloak-js adapter](https://www.npmjs.com/package/keycloak-js) for bearer token based authentication, see [PR 126](https://github.com/terrestris/shogun-demo-client/pull/126) as a reference

Please review @terrestris/devs (and please also note all referenced PRs).

## Related issues or pull requests



--

## Pull request type






## Do you introduce a breaking change?


## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the 
  [Apache Licence Version 2.0](https://github.com/terrestris/shogun/blob/main/LICENSE).
- [x] I have followed the [guidelines for contributing](https://github.com/terrestris/shogun/blob/main/CONTRIBUTING.md).
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/terrestris/shogun/blob/main/CODE_OF_CONDUCT.md).
- [x] I have added or updated tests and documentation, and the test suite passes (run `mvn test` locally).
- [ ] I have added a screenshot/screencast to illustrate the visual output of my update.
